language/bytecode_verifier/src/acquires_list_verifier.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module implements a checker for verifying properties about the acquires list on function
+//! definitions. Function definitions must annotate the global resources (declared in that module)
+//! accesssed by `BorrowGlobal`, `MoveFrom`, and any transative function calls
+//! The list of acquired resources (stored in `FunctionDefinition`'s `acquires_global_resources`
+//! field) must:
+//! - They must contain no duplicate resources (checked by `check_duplication`)
+//! - There must be no missing resources (any resource acquired must be present)
+//! - There must be no additional resources (no extraneous resources not actually acquired)
+
+use crate::control_flow_graph::{BasicBlock, VMControlFlowGraph};
+use std::collections::BTreeSet;
+use vm::{
+    access::ModuleAccess,
+    errors::VMStaticViolation,
+    file_format::{Bytecode, CompiledModule, FunctionDefinition, StructDefinitionIndex},
+    views::{FunctionDefinitionView, ModuleView, StructDefinitionView, ViewInternals},
+};
+
+pub struct AcquiresVerifier<'a> {
+    module_view: ModuleView<'a, CompiledModule>,
+    annotated_acquires: BTreeSet<StructDefinitionIndex>,
+    cfg: &'a VMControlFlowGraph,
+    actual_acquires: BTreeSet<StructDefinitionIndex>,
+    errors: Vec<VMStaticViolation>,
+}
+
+impl<'a> AcquiresVerifier<'a> {
+    pub fn verify(
+        module: &'a CompiledModule,
+        function_definition: &'a FunctionDefinition,
+        cfg: &'a VMControlFlowGraph,
+    ) -> Vec<VMStaticViolation> {
+        let annotated_acquires = function_definition
+            .acquires_global_resources
+            .iter()
+            .cloned()
+            .collect();
+        let mut verifier = Self {
+            module_view: ModuleView::new(module),
+            annotated_acquires,
+            cfg,
+            actual_acquires: BTreeSet::new(),
+            errors: vec![],
+        };
+
+        let function_definition_view = FunctionDefinitionView::new(module, function_definition);
+        let code = &function_definition_view.code().code;
+        for (_, block) in verifier.cfg.blocks.iter() {
+            verifier.verify_block(&block, code)
```

Comment:
I was thinking that while I was writing it. And I was unsure if this would be safe (I was unsure of the layout of the bytecode instructions. Should I do that?